### ISSUE TITLE
Push notification to users in running environments

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/jupyterhub#3279]
## Background

JupyterHub currently support a traitlet configurable announcement to show a banner on pages managed by JupyterHub's web server. But JupyterHub currently doesn't provide a way to notify running servers about something. This could be useful by `jupyterhub-idle-culler` for example but also more generally by custom services developed, for example to give quick feedback to users that starts 1000 workers

## Implementation idea seed

I imagine a technical implementation to support this would be to let JupyterHub listen to a REST API that pre-registered services could make use of, and it would then push a notification to user servers using another well defined REST API to receive such notifications implemented by `jupyter_server` and/or `notebook` for example. As a final step, that this information is presented to the specific UI the user currently works in, such as Jupyter Lab.

To overview this, I think we have a few technical steps involved: a) For JupyterHub to _listen to intent_ to push a notification, b) for JupyterHub to _push a notification_, c) for the user environment to _listen for notifications_, and d) for the user environment to _present notifications_.

## Help

I'd love to _input to evaluate developing a feature like this_: what would a sensible scope and technical implementation strategy be? Ping @jupyterhub/jupyterhubteam and @fperez!

**Comment by manics** [comment]
Do you think some of the [Jupyter RTC work](https://jupyter-rtc.readthedocs.io/en/latest/developer/specs.html) could be used, or modified to handle the messaging?

**Comment by rcthomas** [comment]
https://github.com/rcthomas/jupyterhub-announcement exposes a REST endpoint to grab announcements from.  I've thought of a plug-in on the notebook side that would just hit that endpoint periodically (not a push).

**Comment by betatim** [comment]
A potential rabbit hole is that a JupyterHub can be used to run lots of things like Jupyter notebook, VS Code or RStudio. So the "display a message" part needs to be super flexible so that we have a chance to implement it for these other UIs. I don't know what the right thing to do is but I wanted to surface this.

We have also discussed how to have a button on servers launched by users on mybinder.org (https://github.com/jupyterhub/mybinder.org-deploy/issues/1462 has some thoughts. There are more issues/forum threads which I can try and find). The similarity with this issue is that we need to solve the problem of "how to display something on all user's servers no matter what they are running".

**Comment by rcthomas** [comment]
@betatim I guess apps that the hub launches would need to "opt in" to push notifications.

**Comment by manics** [comment]
> The similarity with this issue is that we need to solve the problem of "how to display something on all user's servers no matter what they are running".

I don't think we can (or should) solve that problem ourselves. Javascript overlays will work for some cases but not all, and in some cases will break a client UI. I like @consideRatio's breakdown of the issue into the four steps:

> a) For JupyterHub to listen to intent to push a notification

That's something fully within our control, could perhaps be prototyped as a service and then moved into core?

> b) for JupyterHub to push a notification
> c) for the user environment to listen for notifications

Are you intending this to be a [Jupyter Enhancement Proposal](https://github.com/jupyter/enhancement-proposals)? Or are you thinking of defining this in some other way?
For prototyping the JupyterHub service plus a JupyterLab extension could be a start?

> d) for the user environment to present notifications.

If this is a JEP then this follows on naturally. It could initially be implemented as a client extensions.

**Comment by betatim** [comment]
I should have made it more explicit that linking the mybinder.org issue was meant as a "we also want this, there is some discussion, but no good solutions yet". At least I don't think we have a good solution for mybinder.org yet.

**Comment by Cyb4Black** [comment]
Did anyone work on this lately? We are currently looking for this exact feature 😅

**Comment by BwL1289** [comment]
> Did anyone work on this lately? We are currently looking for this exact feature 😅

Also looking for this. IMO, it makes sense for this to be part of the hub itself. The ability for the Hub to broadcast messages or notifications to user JupyterLab servers feels like core “Hub responsibility” territory—since the Hub already acts as the central authority, entry point, and pseudo-orchestrator for user sessions.

Having this built into the Hub rather than scattered across external services/extensions would also make it easier to manage, configure, and keep consistent across deployments.
